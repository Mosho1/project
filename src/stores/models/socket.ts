import { types, getParent, hasParent } from 'mobx-state-tree'
import { values } from '../utils/utils';
import { modelTypes } from './index';
import { pouch } from '../utils/pouchdb-model';
import { IStore } from '../domain-state';
import { CodeBlockIO, ICodeBlockIO } from './code-block';
import { IPrimitiveTypes } from './types';

const socketType = types.enumeration('socketType', [
    'input',
    'output',
    'execInput',
    'execOutput',
]);

export type SocketTypeEnum = typeof socketType.Type;

export const socketTypes = (socketType as any).types.map((x: any) => x.value) as SocketTypeEnum[];

export const areSocketsCompatible = (s1: ISocket, s2: ISocket) => {
    if (s1.isInput === s2.isInput) return false;
    if (s1.isExec !== s2.isExec) return false;
    if (s1.code.type !== 'any' &&
        s2.code.type !== 'any' &&
        s1.code.type !== s2.code.type) return false;
    const [input, output] = s1.isInput ? [s1, s2] : [s2, s1];
    if (input.isExec && (input.arrows.length > 0 || output.arrows.length > 0)) return false;
    if (input.arrows.length > 0) return false;
    return true;
};

const typeColors: {[T in IPrimitiveTypes]: string} = {
    any: '#fff',
    number: 'green',
    string: 'red',
    boolean: '#1565C0',
    void: '#000'
};

export const Socket = pouch.model('Socket', {
    socketType,
    name: types.optional(types.string, ''),
    code: types.reference<ICodeBlockIO>(CodeBlockIO)
}).volatile(_self => ({
    value: null,
}))
    .views(self => ({
        get store(): null | IStore {
            if (!hasParent(self, 2)) return null;
            return getParent(self, 2);
        },
    }))
    .views(self => ({
        get box(): null | modelTypes['Box'] {
            if (!self.store) return null;
            return values(self.store.boxes).find(b => Boolean(b.sockets.find(s => s === self))) || null;
        },
    }))
    .views(self => ({
        get index() {
            if (!self.box) return 0;
            const { inputs, outputs, execInputs, execOutputs } = self.box;
            switch (self.socketType) {
                case 'input':
                    return execInputs.length + inputs.findIndex(s => s === self);
                case 'output':
                    return execOutputs.length + outputs.findIndex(s => s === self);
                case 'execInput':
                    return execInputs.findIndex(s => s === self);
                case 'execOutput':
                    return execOutputs.findIndex(s => s === self);
            }
            /* istanbul ignore next */
            return 0;
        },
    }))
    .views(self => ({
        get isExec() {
            return self.socketType === 'execInput' || self.socketType === 'execOutput';
        },
        get isInput() {
            return self.socketType === 'execInput' || self.socketType === 'input';
        },
        get arrows() {
            if (!self.box || !self.box.store) return [];
            return values(self.box.store.arrows).filter(a => a.input === self || a.output === self);
        },
        get x() {
            if (!self.box) return 0;
            if (self.socketType === 'input' || self.socketType === 'execInput') {
                return 18;
            } else {
                return self.box.width - 18;
            }
        },
        get y() {
            return 50 + self.index * 30;
        },

    }))
    .views(self => ({
        get absX() {
            if (!self.box) return 0;
            return self.x + self.box.x;
        },
        get absY() {
            if (!self.box) return 0;
            return self.y + self.box.y;
        },
        get color() {
            return typeColors[self.code.type] || '#fff';
        },
    }))
    .views(self => ({
        get fillColor() {
            switch (self.code.type) {
                case 'any':
                    if (self.arrows.length > 0) {
                        return self.arrows[0].color;
                    }
                default:
                    return self.color;
            }
        }
    }))
    .actions(self => {
        return {
            setName(name: string | null) {
                self.name = name || '';
            },
            isCompatibleWith(socket: typeof self) {
                return areSocketsCompatible(self as any, socket as any);
            },
            setValue(value: any) {
                /* istanbul ignore next */
                self.value = value;
            }
        }
    });

type ISocketType = typeof Socket.Type;
type ISocketSnapshotType = typeof Socket.SnapshotType;
export interface ISocket extends ISocketType { };
export interface ISocketSnapshot extends ISocketSnapshotType { };