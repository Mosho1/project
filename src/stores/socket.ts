// eslint-disable-next-line
import {
    onSnapshot,
    applySnapshot,
} from "mobx-state-tree"

let subscription: any;
export default function syncStoreWithBackend(socket: any, store: any) {
    // === SYNC PATCHES (recommended)
    // subscription = onPatch(store, (data) => {
    //     socketSend(data)
    // })

    // onSocketMessage((data) => {
    //     applyPatch(store, data)
    // })

    // === SYNC ACTIONS
    // subscription = onAction(store, data => {
    //     socketSend(data)
    // })

    // onSocketMessage(data => {
    //     applyAction(store, data)
    // })

    // === SYNC SNAPSNOTS
    subscription = onSnapshot(store, (data) => {
        socketSend(data)
    })

    onSocketMessage((data: any) => {
        applySnapshot(store, data)
    })

    let isHandlingMessage = false
    function socketSend(data: any) {
        if (!isHandlingMessage) socket.send(JSON.stringify(data))
    }

    function onSocketMessage(handler: any) {
        socket.onmessage = (event: any) => {
            isHandlingMessage = true
            handler(JSON.parse(event.data))
            isHandlingMessage = false
        }
    }
}

/**
 * Clean up old subscription when switching communication system
 */
if (module.hot) {
    module.hot.dispose(() => {
        subscription();
    })
}
