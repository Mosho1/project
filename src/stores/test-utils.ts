import { mapValues } from 'lodash';

export const mock = <T>(instance: T, mock: NestedPartial<T>) => {
    for (let k in mock) {
        const descriptor = Object.getOwnPropertyDescriptor(instance, k);
        Object.defineProperties(instance, mapValues(mock, value => ({
            value,
            enumerable: descriptor!.enumerable,
            configurable: true,
            writable: true,
        })));
    }
    return instance;
};

const f = (a: any[], b: any[]): any[] =>
    [].concat(...a.map(a2 => b.map((b2: any) => [].concat(a2 as any, b2 as any) as any) as any));

export const product = (a: any[], b: any[], ...c: any[]): any[][] => {
    if (!b || b.length === 0) {
        return a;
    }
    const [b2, ...c2] = c;
    const fab = f(a, b);
    return product(fab, b2, ...c2);
};