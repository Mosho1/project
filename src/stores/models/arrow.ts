import { types } from 'mobx-state-tree'
import { Socket } from './socket';
import { calculateBezierPoints } from '../utils/utils';
import { pouch } from '../utils/pouchdb-model';



export const Arrow = pouch.model('Arrow', {
    input: types.reference(Socket),
    output: types.reference(Socket)
}).views(self => {
    return {
        get points() {
            return calculateBezierPoints(
                self.output.x,
                self.output.y,
                self.input.x,
                self.input.y,
            );
        },
        get isExec() {
            return self.input.isExec;
        }
    };
});
