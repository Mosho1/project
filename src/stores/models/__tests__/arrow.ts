import { Arrow } from "../arrow";
import { Socket } from '../socket';
import { mock } from '../../test-utils';


test('points', () => {
    const input = mock(Socket.create({ socketType: 'input' }), { box: null });
    const output = mock(Socket.create({ socketType: 'output' }), { box: null });
    const a = Arrow.create({ input, output });
    expect(a.points).toMatchSnapshot();
});

describe('isExec', () => {
    test('false', () => {
        const input = mock(Socket.create({ socketType: 'input' }), { box: null });
        const output = mock(Socket.create({ socketType: 'output' }), { box: null });
        const a = Arrow.create({ input, output });
        expect(a.isExec).toBe(false);
    });

    test('true', () => {
        const input = mock(Socket.create({ socketType: 'execInput' }), { box: null });;
        const output = mock(Socket.create({ socketType: 'execOutput' }), { box: null });
        const a = Arrow.create({ input, output });
        expect(a.isExec).toBe(true);
    });
});
