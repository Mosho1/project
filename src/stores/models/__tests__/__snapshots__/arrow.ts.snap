// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`points 1`] = `
Array [
  0,
  0,
  50,
  0,
  -50,
  0,
  0,
  0,
]
`;
